### ISSUE TITLE
transformer adam first-step param explosion (4.7× L2 jump in one step) + post-explosion loss oscillation

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1266]
## Symptom

Direct trace of one \`Transformer.Train()\` step against PR #1265's patched source (Adam default, LR=1e-3, V=4 / ctxLen=8 / d_model=16 / d_ff=32 / 1 layer / 2 heads):

\`\`\`
Default optimizer:    AdamOptimizer\`3
Optimizer options:    MaxIterations=100, InitialLR=0.001, UseAdaptiveLR=True, LRDecay=0.99
Total param L2 BEFORE training: 2.365063
Initial logits:       [0.0556, 0.9231, 0.0049, 0.0164]
Initial probs:        [0.1889, 0.4498, 0.1796, 0.1817]
Loss after 1 step:    0.087436
Total param L2 AFTER 1 step:  11.207761  (Δ = +8.842698)
Loss after 100 steps: 0.092077
\`\`\`

Two distinct bugs:

1. **Param L2 jumps 4.7× in ONE step** (2.37 → 11.21). Per-parameter update magnitude ≈ 0.13 (= 8.85 / √5000 params), which is **125× the configured LR=1e-3**. Adam at default β values should produce update magnitude ≤ LR per parameter on the first step (≈ sign(grad) × LR), not 125× LR.
2. **Loss INCREASES from step 1 to step 100** (0.087 → 0.092) — the model is oscillating around the post-explosion bad point instead of converging. Even after 99 more steps the optimizer can't recover from where the first step blew it.

## Suspected location

\`NeuralNetworkBase.TrainWithTape\` line 3506:
\`\`\`csharp
if (TryTrainWithFusedOptimizer(input, expected, resolvedOptimizer))
    return;
\`\`\`

The fused-compiled fast path runs first for plain Adam/AdamW/SGD with constant hyperparameters and \`T = float\` (Tensor 0.71 GEMM kernels operate on \`float*\` directly). Transformer with default Adam at float matches all preconditions, so this path engages on every step.

Hypothesis: the fused kernel applies Adam's first-step bias correction (\`m_hat = m / (1-β₁¹) = m × 10\`, \`v_hat = v / (1-β₂¹) = v × 1000\`) AND something else — possibly a missing initialization of the m/v moment buffers, OR a per-tensor scale factor that doesn't account for fp32 init magnitude. Without the trace I can't tell whether the bug is in:
- AiDotNet.Tensors-side fused-optimizer kernel
- AiDotNetEngine binding's argument marshaling  
- AiDotNet's TryTrainWithFusedOptimizer pre-compile setup

If \`TryTrainWithFusedOptimizer\` falls back to the eager path (\`return false\`), the eager Adam.Step at \`AdamOptimizer.cs:482\` produces normal-magnitude updates per the standard formula. So the explosion is specifically in the fused path.

## Diagnostic to confirm

Disable the fused path for Adam at line 3506 and rerun the trace. If param L2 stays bounded, the bug is in the fused-optimizer kernel. If it still explodes, look further upstream (TapeStepContext setup, paramBuffer alignment, etc.).

## Repro

Test added to PR #1265 at \`tests/AiDotNet.Tests/IntegrationTests/NeuralNetworks/TransformerTrainingTraceTest.cs\`. Run:
\`\`\`
dotnet test --filter \"FullyQualifiedName~TransformerTrainingTraceTest\" --logger \"console;verbosity=detailed\"
\`\`\`

Strong invariants \`OptimizerStep_ParamL2_DoesNotExplode\` and \`LossStrictlyDecreasesOnMemorizationTask\` added to \`NeuralNetworkModelTestBase\` in the same PR will fail loud once auto-applied to Transformer's generated test class via TestScaffoldGenerator — that's the acceptance bar for this fix.

## Severity

Critical for Transformer training. Combined with the 4 fixes already in PR #1265 (Adam default, OptimizerBase counter inversion, TrainBatched, AiModelBuilder facade rewire), all upstream training-pipeline bugs are visible to invariant tests but actual convergence still requires this last explosion fix.

**Comment by ooples** [comment]
False alarm — diagnostic was wrong. The 'param explosion' was a lazy-init artifact in the test (LayerNorm γ initializes to 1.0 only on first forward pass; my BEFORE measurement undercounted). With a warmup forward pass before L2 measurement, the trace test passes. Adam first-step update is correct (~√N × LR magnitude). Fix committed in PR #1265 (commit on top of 332f41edf). The convergence rate at byte-LM scale (#1264) is now the only remaining work and that's a 'mathematically-slow per-sample updates' framing not a bug.
